test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java

```diff
@@ -1478,18 +1478,25 @@ private void postIndexAsyncActions(String[] indices, List<CountDownLatch> inFlig
         int numDataNodes() default -1;
 
         /**
-         * Returns the minimum number of nodes in the cluster. Default is <tt>-1</tt>.
+         * Returns the minimum number of data nodes in the cluster. Default is <tt>-1</tt>.
          * Ignored when {@link ClusterScope#numDataNodes()} is set.
          */
         int minNumDataNodes() default -1;
 
         /**
-         * Returns the maximum number of nodes in the cluster.  Default is <tt>-1</tt>.
+         * Returns the maximum number of data nodes in the cluster.  Default is <tt>-1</tt>.
          * Ignored when {@link ClusterScope#numDataNodes()} is set.
          */
         int maxNumDataNodes() default -1;
 
         /**
+         * Returns the number of master nodes in the cluster. <tt>0</tt> means data nodes will sever as master nodes
+         * and there will be no dedicated master (and data) nodes. Default is <tt>-1</tt> which means
+         * a random number of nodes is used, potentially 0 (meaning data nodes will assume master role)
+         */
+        int numMasterNodes() default -1;
```

Comment:
The tricky part here is that if we do this, we have inconsistencies with the numDataNodes attribute. We can't really call that numDataOnlyNodes because of the fact that they might be used as master only nodes. I agree with you that this is all note ideal as you commented in other places, but I don't have any good way to make it better. I'm very open to suggestions.